File: 2025-09/2025-09-07/romanNumeralParser.js
function parseRomanNumeral(numeral) {
  const romanNumerals = {
    I : 1,
    V : 5,
    X : 10,
    L : 50,
    C : 100,
    D : 500,
    M : 1000
  };
  let result = 0;

  for (let i = 0; i < numeral.length - 1; i ++) {
    if (romanNumerals[numeral[i]] < romanNumerals[numeral[i + 1]])
      result -= romanNumerals[numeral[i]];
    else
      result += romanNumerals[numeral[i]];
  }
  result += romanNumerals[numeral[numeral.length - 1]];

  return result;
}

console.log(parseRomanNumeral("III"));
console.log(parseRomanNumeral("IV"));
console.log(parseRomanNumeral("XXVI"));
console.log(parseRomanNumeral("XCIX"));
console.log(parseRomanNumeral("CDLX"));
console.log(parseRomanNumeral("DIV"));
console.log(parseRomanNumeral("MMXXV"));

console.log(parseRomanNumeral("I"));

// --- TEST SUITE ---

const testsText = `
1. parseRomanNumeral("III") should return 3.
2. parseRomanNumeral("IV") should return 4.
3. parseRomanNumeral("XXVI") should return 26.
4. parseRomanNumeral("XCIX") should return 99.
5. parseRomanNumeral("CDLX") should return 460.
6. parseRomanNumeral("DIV") should return 504.
7. parseRomanNumeral("MMXXV") should return 2025.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);

File: 2025-09/2025-09-07/roman_numeral_parser.py
def parse_roman_numeral(numeral):
    roman_numerals = {
        'I' : 1,
        'V' : 5,
        'X' : 10,
        'L' : 50,
        'C' : 100,
        'D' : 500,
        'M' : 1000
    }
    result = 0

    for i in range(len(numeral) - 1):
        if roman_numerals[numeral[i]] < roman_numerals[numeral[i + 1]]:
            result -= roman_numerals[numeral[i]]
        else:
            result += roman_numerals[numeral[i]]
    
    result += roman_numerals[numeral[-1]]

    return result

print(parse_roman_numeral("III"))
print(parse_roman_numeral("IV"))
print(parse_roman_numeral("XXVI"))
print(parse_roman_numeral("XCIX"))
print(parse_roman_numeral("CDLX"))
print(parse_roman_numeral("DIV"))
print(parse_roman_numeral("MMXXV"))

print(parse_roman_numeral("I"))

# --- TEST SUITE ---

tests_text = '''
1. parse_roman_numeral("III") should return 3.
2. parse_roman_numeral("IV") should return 4.
3. parse_roman_numeral("XXVI") should return 26.
4. parse_roman_numeral("XCIX") should return 99.
5. parse_roman_numeral("CDLX") should return 460.
6. parse_roman_numeral("DIV") should return 504.
7. parse_roman_numeral("MMXXV") should return 2025.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)